File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-4.java
class Solution {
    public int search(int[] nums, int target) {
        int low = 0;
        int high = nums.length - 1;
        
        while(low <= high){
            int mid = (low + high) / 2;
            System.out.println("low: "+low);
            System.out.println("high: "+high);
            System.out.println("mid: "+mid);
            if(nums[mid] == target){
                return mid;
            }else if(nums[low] < nums[mid] && target < nums[mid] && target >= nums[low]){
                high = mid - 1;
            }else if(nums[low] < nums[mid] && !(target < nums[mid] && target >= nums[low])){
                low = mid + 1;
            }else if(nums[mid] < nums[high] && target > nums[mid] && target <= nums[high]){
                low = mid + 1; 
            }else if(nums[mid] < nums[high] && !(target > nums[mid] && target <= nums[high])){
                high = mid - 1;
            }else{
                // System.out.println("I m here");
                low = mid + 1;
            }
        }
        return -1;
    }
}

File: Data Structures & Algorithms/search-2d-matrix/submission-4.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        rows = len(matrix)
        cols = len(matrix[0])

        low = 0
        high = self.indicesToIndex(cols, rows - 1, cols - 1)

        while low <= high:
            mid = (low + high) // 2
            midIndices = self.indexToIndices(cols, mid)
            if matrix[midIndices[0]][midIndices[1]] == target:
                return True
            elif  matrix[midIndices[0]][midIndices[1]] > target:
                high = mid - 1
            else:
                low = mid + 1
        return False               





    def indicesToIndex(self, cols: int, x: int, y: int) -> int:
        return x * cols + y 


    def indexToIndices(self, cols: int, ind: int) -> Tuple[int, int]:
        row = ind//cols             
        return (row, (ind - (cols * row))) 